server: Add sampling rate to distributed tracing

lilic opened the pull request:
In the initial PR we mentioned we should also add sampling rate, as this will allow users to control the overhead by being able to reduce the amount of traces sent.

This PR adds `ExperimentalDistributedTracingSamplingRatePerMillion`, which is the number of samples to collect per million spans.

It defaults to 0, I wasn't sure if we should default to sample all instead of 0, but this is how the Kubernetes implementation did it so I tried to follow the same principles as there.

serathius commented on server/embed/config_tracing.go:
```diff
@@ -0,0 +1,110 @@
+
+const maxSamplingRatePerMillion = 1000000
+
+func (e *Etcd) setupTracingExporter(ctx context.Context) (exporter tracesdk.SpanExporter, options []otelgrpc.Option, err error) {
```
Can change this to method that takes ServerConfig as argument. This function doesn't need anything from embed.Etcd than e.cfg. As so we should try to keep is separate from embed.Etcd [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter) and make them separate. 

Best even move it to separate package `server/embed/tracing` this way we can avoid this function reading embed.Etcd private fields and functions.

lilic replied:
I followed the pattern we have with logging here, where we have config_logging.go. But will change that takes config instead, good suggestion!

serathius replied:
Ok, good point. I also prefer consistent quality, instead of having each module written differently. 
But this is a good candiate for a follow up PR. I would be happy to help with that.

serathius commented on server/embed/config.go:
```diff
@@ -348,6 +348,9 @@ type Config struct {
 	// Can only be used if ExperimentalEnableDistributedTracing is true.
 	ExperimentalDistributedTracingServiceInstanceID string `json:"experimental-distributed-tracing-instance-id"`
+	// ExperimentalDistributedTracingSamplingRatePerMillion is the number of samples to collect per million spans.
+	// Defaults to 0.
```
I don't think `0` is sensible default. Please provide a value that makes sense when someone enabled tracing for the first time. 

I also think `0` should be forbidden as this is an invalid value, we should not use it to disable tracing as we have a separate flag for it `"experimental-enable-distributed-tracing"`. Configuration flags should be perpendicular (2 flags should not be responsible for enabling/disabling the feature), as so we should not allow setting `experimental-distributed-tracing-sampling-rate` to zero to disable tracing, as users should use `experimental-enable-distributed-tracing` for that.

serathius commented on server/embed/config_tracing.go:
```diff
@@ -0,0 +1,110 @@
+	if samplingRate == 0 {
+		return sampler, nil
+	}
+	if samplingRate < 0 {
```
This looks like flag value validation, please move it to embed.Config.Validate

lilic replied:
Good suggestion, for both this and the below check right?

serathius replied:
I think so. Best would be to have a separate configuration struct for tracing that is included in embed.Config. This way we can have a validate function in tracing package that is invoked by embed.Config.Validate (common pattern in K8s code).

lilic replied:
yeah agreed its bit different per configuration here. The cfg.setupLogging seems to be called in validate as well. But agreed we can do it down the road in similar way, especially since config is really growing and lacks testing IMO.

lilic commented:
@serathius let me know what you think, I addressed the suggestions, thanks. Will move some of unit tests cases if this is the approach you thought of.

serathius commented on server/embed/config_tracing.go:
```diff
@@ -0,0 +1,115 @@
+	return exporter, options, err
+}
+
+func determineSampler(samplingRate int) (tracesdk.Sampler, error) {
```
This always returns nil for error

lilic replied:
nice catch, removed error returns.

serathius reviewed (approved):
Looks good aside one small nit.

lilic commented:
Thanks, separated the unit tests for new function and fixed the nit. Please have another look, thanks!

hexfusion commented on server/embed/config_tracing.go:
```diff
@@ -0,0 +1,112 @@
+// never use the empty default string value, it's set if
+// if it's a non empty string.
+func determineResourceWithIDKey(serviceInstanceID string) *resource.Resource {
+	if serviceInstanceID == "" {
```
optional: have function return nil
```
if serviceInstanceID != "" {
```

hexfusion commented on server/etcdmain/help.go:
```diff
@@ -210,6 +210,8 @@ Experimental distributed tracing:
     Distributed tracing service name, must be same across all etcd instances.
   --experimental-distributed-tracing-instance-id ''
     Distributed tracing instance ID, must be unique per each etcd instance.
+  --experimental-distributed-tracing-sampling-rate '0'
```
optional: worth identifying 0 as disabled?

lilic replied:
> It defaults to 0, I wasn't sure if we should default to sample all instead of 0, but this is how the Kubernetes implementation did it so I tried to follow the same principles as there.

Yeah, this was my main question in the description, as this disables it essentially not sure but I guess you are right if we document that this makes it disabled than we should maybe not make it optional? 🤔

hexfusion commented on server/embed/etcd.go:
```diff
@@ -238,6 +231,10 @@ func StartEtcd(inCfg *Config) (e *Etcd, err error) {
 		srvcfg.ExperimentalTracerOptions = opts
+
+		e.cfg.logger.Info(
+			"distributed tracing started",
```
optional: are we started or enabled here?

lilic replied:
Depends on the sampling I guess, we register it maybe I can change it to that?

hexfusion reviewed (approved):
few optional notes, otherwise LGTM

hexfusion commented on server/etcdmain/help.go:
```diff
@@ -210,6 +210,8 @@ Experimental distributed tracing:
   --experimental-distributed-tracing-instance-id ''
     Distributed tracing instance ID, must be unique per each etcd instance.
+  --experimental-distributed-tracing-sampling-rate '0'
+	Number of samples to collect per million spans for distributed tracing. Disabled by default.
```
minor: indent spacing seems off

lilic replied:
nice catch, done!

hexfusion reviewed (approved):
one very minor nit then lets merge :+1:

hexfusion commented:
Proxy failure appears to be a flake and passed on the previous run (txt change diff)
